-1

Tested at: 7aa79a9c9630f6afa92fc15603d5ab9ff529552f

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1786a3/1311/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/CompositeCandidate.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/TriggerPath.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/Conversion.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/throwMissingLabel.cc 
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/Jet.cc: In member function 'void pat::Jet::scaleEnergy(double, std::__cxx11::string)':
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/Jet.cc:250:111: error: expected '}' before ')' token
     std::vector<float> factors = { jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_)*fScale) };
                                                                                                               ^
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/Jet.cc:250:104: error: narrowing conversion of '(((double)(&((pat::Jet*)this)->pat::Jet::jec_.std::vector<pat::JetCorrFactors>::operator[](((std::vector<pat::JetCorrFactors>::size_type)((pat::Jet*)this)->pat::Jet::currentJECSet_)))->pat::JetCorrFactors::correction(((pat::Jet*)this)->pat::Jet::currentJECLevel_, ((pat::Jet*)this)->pat::Jet::currentJECFlavor_)) * fScale)' from 'double' to 'float' inside { } [-Werror=narrowing]
     std::vector<float> factors = { jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_)*fScale) };
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
</pre>

* **Clang**:

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/TauJetCorrFactors.cc 
>> Building LCG reflex dict from header file src/DataFormats/PatCandidates/src/classes_objects.h
>> Building LCG reflex dict from header file src/DataFormats/PatCandidates/src/classes_user.h
>> Building LCG reflex dict from header file src/DataFormats/PatCandidates/src/classes_trigger.h
>> Building LCG reflex dict from header file src/DataFormats/PatCandidates/src/classes_other.h
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/Jet.cc:250:111: error: expected '}'
    std::vector<float> factors = { jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_)*fScale) };
                                                                                                              ^
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-07-03-1100/src/DataFormats/PatCandidates/src/Jet.cc:250:34: note: to match this '{'
    std::vector<float> factors = { jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_)*fScale) };
                                 ^
</pre>

--- a/DataFormats/PatCandidates/interface/Jet.h
+++ b/DataFormats/PatCandidates/interface/Jet.h
@@ -175,7 +175,7 @@ namespace pat {
     };
     /// Scale energy and correspondingly adjust raw jec factors
     void scaleEnergy(double fScale) override { scaleEnergy(fScale, "scaled"); };
-    void scaleEnergy(double fScale, std::string level);
+    void scaleEnergy(double fScale, const std::string& level);
 
   private:
     /// index of the set of jec factors with given label; returns -1 if no set

--- a/DataFormats/PatCandidates/interface/JetCorrFactors.h
+++ b/DataFormats/PatCandidates/interface/JetCorrFactors.h
@@ -35,8 +35,6 @@
 namespace pat {
 
   class JetCorrFactors {
-    friend class Jet;
-
   public:
     // jet energy correction factor. For flavor independent jet energy corrections the
     // std::vector<float> holds just a single entry. From the first flavor dependent entry
@@ -54,6 +52,8 @@ namespace pat {
     JetCorrFactors(){};
     // constructor by value
     JetCorrFactors(const std::string& label, const std::vector<CorrectionFactor>& jec);
+    // add correction factor
+    void addFactor(const CorrectionFactor& factor);
 
     // instance label of the jet energy corrections set
     std::string jecSet() const { return label_; }
@@ -92,6 +92,7 @@ namespace pat {
     bool flavorIndependent(const CorrectionFactor& jec) const { return (jec.second.size() == 1); }
     // check consistency of input vector
     bool isValid(const CorrectionFactor& jec) const { return (flavorDependent(jec) || flavorIndependent(jec)); }
+    void invalidFactor() const;
 
   private:
     // instance label of jet energy correction factors

--- a/DataFormats/PatCandidates/src/Jet.cc
+++ b/DataFormats/PatCandidates/src/Jet.cc
@@ -247,7 +247,7 @@ const reco::JetFlavourInfo& Jet::jetFlavourInfo() const { return jetFlavourInfo_
 /// Scale energy and correspondingly add jec factor
 void Jet::scaleEnergy(double fScale, std::string level) {
   if (jecSetsAvailable()) {
-    std::vector<float> factors = { jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_)*fScale) };
+    std::vector<float> factors = {float(jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_) * fScale)};
     jec_[currentJECSet_].jec_.push_back(std::make_pair(level, factors));
     currentJECLevel(jec_[currentJECSet_].jec_.size() - 1);
     currentJECFlavor(JetCorrFactors::NONE);

--- a/DataFormats/PatCandidates/src/Jet.cc
+++ b/DataFormats/PatCandidates/src/Jet.cc
@@ -245,7 +245,7 @@ const reco::JetFlavourInfo& Jet::jetFlavourInfo() const { return jetFlavourInfo_
 /// ============= Jet Energy Correction methods ============
 
 /// Scale energy and correspondingly add jec factor
-void Jet::scaleEnergy(double fScale, std::string level) {
+void Jet::scaleEnergy(double fScale, const std::string& level) {
   if (jecSetsAvailable()) {
     std::vector<float> factors = {float(jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_) * fScale)};
     jec_[currentJECSet_].jec_.push_back(std::make_pair(level, factors));

--- a/DataFormats/PatCandidates/src/Jet.cc
+++ b/DataFormats/PatCandidates/src/Jet.cc
@@ -248,8 +248,8 @@ const reco::JetFlavourInfo& Jet::jetFlavourInfo() const { return jetFlavourInfo_
 void Jet::scaleEnergy(double fScale, const std::string& level) {
   if (jecSetsAvailable()) {
     std::vector<float> factors = {float(jec_[currentJECSet_].correction(currentJECLevel_, currentJECFlavor_) * fScale)};
-    jec_[currentJECSet_].jec_.push_back(std::make_pair(level, factors));
-    currentJECLevel(jec_[currentJECSet_].jec_.size() - 1);
+    jec_[currentJECSet_].addFactor(std::make_pair(level, factors));
+    currentJECLevel(jec_[currentJECSet_].numberOfCorrectionLevels() - 1);
     currentJECFlavor(JetCorrFactors::NONE);
   }
   setP4(p4() * fScale);

--- a/DataFormats/PatCandidates/src/JetCorrFactors.cc
+++ b/DataFormats/PatCandidates/src/JetCorrFactors.cc
@@ -12,29 +12,15 @@ using namespace pat;
 JetCorrFactors::JetCorrFactors(const std::string& label, const std::vector<CorrectionFactor>& jec)
     : label_(label), jec_(jec) {
   for (std::vector<CorrectionFactor>::const_iterator corrFactor = jec.begin(); corrFactor != jec.end(); ++corrFactor) {
-    if (!isValid(*corrFactor)) {
-      throw cms::Exception("InvalidRequest")
-          << "You try to create a CorrectionFactor which is neither flavor dependent nor \n"
-          << "flavor independent. The CorrectionFactor should obey the following rules:  \n"
-          << "\n"
-          << " * CorrectionFactor is a std::pair<std::string, std::vector<float> >.      \n"
-          << " * The std::string holds the label of the correction level (following the  \n"
-          << "   conventions of JetMET.                                                  \n"
-          << " * The std::vector<float> holds the correction factors, these factors are  \n"
-          << "   up to the given level. They include all previous correction steps.      \n"
-          << " * The vector has the length *1* for flavor independent correction factors \n"
-          << "   or *5* for flavor dependent correction factors.                         \n"
-          << " * The expected order of flavor dependent correction factors is: NONE,     \n"
-          << "   GLUON, UDS, CHARM, BOTTOM. If follows the JetMET conventions and is     \n"
-          << "   in the Flavor enumerator of the JetCorrFactos class.                    \n"
-          << " * For flavor depdendent correction factors the first entry in the vector  \n"
-          << "   (corresponding to NONE) is invalid and should be set to -1. It will not \n"
-          << "   be considered by the class structure though.                            \n"
-          << "\n"
-          << "Make sure that all elements of the argument vector to this contructor are  \n"
-          << "in accordance with these rules.\n";
-    }
+    if (!isValid(*corrFactor))
+      invalidFactor();
   }
+}
+
+void JetCorrFactors::addFactor(const CorrectionFactor& corrFactor) {
+  if (!isValid(corrFactor))
+    invalidFactor();
+  jec_.push_back(corrFactor);
 }
 
 std::string JetCorrFactors::jecFlavor(const Flavor& flavor) const {
@@ -123,4 +109,27 @@ void JetCorrFactors::print() const {
     }
     message << "\n";
   }
+}
+
+void JetCorrFactors::invalidFactor() const {
+  throw cms::Exception("InvalidRequest")
+      << "You try to create a CorrectionFactor which is neither flavor dependent nor \n"
+      << "flavor independent. The CorrectionFactor should obey the following rules:  \n"
+      << "\n"
+      << " * CorrectionFactor is a std::pair<std::string, std::vector<float> >.      \n"
+      << " * The std::string holds the label of the correction level (following the  \n"
+      << "   conventions of JetMET.						      \n"
+      << " * The std::vector<float> holds the correction factors, these factors are  \n"
+      << "   up to the given level. They include all previous correction steps.      \n"
+      << " * The vector has the length *1* for flavor independent correction factors \n"
+      << "   or *5* for flavor dependent correction factors.			      \n"
+      << " * The expected order of flavor dependent correction factors is: NONE,     \n"
+      << "   GLUON, UDS, CHARM, BOTTOM. If follows the JetMET conventions and is     \n"
+      << "   in the Flavor enumerator of the JetCorrFactos class.		      \n"
+      << " * For flavor depdendent correction factors the first entry in the vector  \n"
+      << "   (corresponding to NONE) is invalid and should be set to -1. It will not \n"
+      << "   be considered by the class structure though.			      \n"
+      << "\n"
+      << "Make sure that all elements of the argument vector to this contructor are  \n"
+      << "in accordance with these rules.\n";
 }

